Changelog/Revision History page

nishanthvijayan opened the pull request:
Sets up a Revision History page.
 - The page displays changes to conferences and associated resources.
 - Only organizers can view and revert the revision history of their conferences
 -  Admins can view and revert the revision history for all conferences
 - When describing a change, attributes whose values changed from one type of blank value to another (eg: nil to '  ')   will be ignored
 - The previous and current values of each modified attribute can be viewed by clicking the 'View Changes' button of the corresponding change
 - Revert feature:
  - Reverting the creation of an object deletes it
  - Reverting the deletion of an object re-creates it
  - Reverting the creation of a conference and user is not permitted
  - If a change include changes to multiple attributes, each of these attributes can be reverted individually. But this is allowed only for changes where the object was updated and not created/deleted and if the object is alive (not deleted).
 - Few models have not been tracked as they are expected to be reworked soon (TicketPurchase, Question related stuff)
 - This feature adds a new field named conference_id to the versions table.This is used to fetch versions belonging to a conference/ models belonging to a conference.A rake task has been provided to fill this new column for pre-existing versions (Conference and Event model were already being tracked)
The take task can be invoked by `bundle exec rake data:set_conference_in_versions`

nishanthvijayan commented:
![screen](https://cloud.githubusercontent.com/assets/7537349/15657696/d06f005e-26d0-11e6-9639-af9f22483c75.png)

differentreality commented:
Can you upload a screenshot with pagination please?

differentreality commented on app/models/program.rb:
```diff
@@ -49,7 +51,7 @@ def confirmed
   validates :rating, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 10 }
 
-  before_create :create_event_types
+  after_create :create_event_types
```
This change is not directly relevant to this PR. Could you extract it (along with the changes in factories) into a different PR?

nishanthvijayan replied:
Because of versioning, on trying to create any factory that needs program, it'll throw up an error saying the factory is invalid.
The reason is that on creating a program this before_create creates  2 event_types without a program id.This triggers paper_trail to record the create event of these EventTypes  resulting in a call to the EventType method conference_id which throws up `undefined method conference_id for nil:NilClass (NoMethodError)` as these 2 event types do not have a program at that point.
I included these fixes as the errors are  direct consequences of the changes introduced in this PR.

nishanthvijayan replied:
@differentreality On creating a conference, it creates these 2 event types as well.Do we need to show that in the changelog page.If not paper_trail tracking  of Event Type should be turned off inside `create_event_types`.What do you think?

differentreality replied:
I think the creation of an event type is an action that happens, so it should be logged, even though it is supposed to happen automatically for every conference.

differentreality replied:
What about the next line? Don't we have the same issue with difficulty levels?
https://github.com/openSUSE/osem/blob/master/app/models/program.rb#L53

nishanthvijayan replied:
I wasn't tracking DifficultyLevel.I tried it now locally and it has the same issue.So if were to similarly track DifficultyLevel similar change should be made for [https://github.com/openSUSE/osem/blob/master/app/models/program.rb#L53]( https://github.com/openSUSE/osem/blob/master/app/models/program.rb#L53)

nishanthvijayan commented:
![pagination](https://cloud.githubusercontent.com/assets/7537349/15670091/35b921f4-2741-11e6-9918-667fcfa5b767.png)

hennevogel commented on app/models/role.rb:
```diff
@@ -1,6 +1,9 @@
   belongs_to :resource, polymorphic: true
   has_and_belongs_to_many :users, join_table: :users_roles
+
+  has_paper_trail on: [:update], only: [:name, :description], meta: {conference_id: :resource_id}
```
There is no interface to create roles, why track it?

differentreality replied:
It is only tracking **update** action. Action **toggle_user** will also be tracked.

nishanthvijayan replied:
Actually this only tracks change of description of role.
Tracking adding/removing users (toggle_user) is a bit complicated and I'll add that in another PR.

hennevogel commented on app/views/admin/versions/_event_change_description.html.haml:
```diff
@@ -0,0 +1,35 @@
+- if version.event == 'create'
```
Please use [case](http://www.blackbytes.info/2015/10/ruby-case/)

nishanthvijayan replied:
Which form should I use?
```
case 
  when boolean condition
    statements
  when boolean condition
     statements
```
form?
Or using procs/lambdas?

hennevogel replied:
I use the simple form and would move this into a helper

nishanthvijayan replied:
Okay I'm a bit confused.How do I use the simple form here? the conditions are complex and is not just a simple variable comparison

hennevogel replied:
Use a Boolean case. 

```ruby
irb(main):012:0> result = case
irb(main):013:1* when false && false then 'false'
irb(main):014:1> when false && true then 'half false'
irb(main):015:1> when true && true then 'truthy'
irb(main):016:1> end
=> "truthy"
```

nishanthvijayan replied:
RuboCop offence [EmptyCaseCondition](http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/EmptyCaseCondition)

hennevogel replied:
Please use complete sentences when communicating. It helps understanding what you are trying to say very much :-)

nishanthvijayan replied:
Sorry :) .
Trying to use case like you suggested threw up this [RuboCop offence](http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/EmptyCaseCondition)

hennevogel commented on app/views/admin/versions/_object_desc_and_link.html.haml:
```diff
@@ -0,0 +1,72 @@
+- if !version.item_type.constantize.exists?(version.item_id)
```
Please use [case](http://www.blackbytes.info/2015/10/ruby-case/)

nishanthvijayan commented on app/views/admin/versions/_event_change_description.html.haml:
```diff
@@ -0,0 +1,35 @@
+  = 'submitted new'
+
+- elsif version.event == 'destroy'
+  = 'deleted'
```
@differentreality  There is no interface to delete an event.Should I remove this?

differentreality replied:
Yeap you should

nishanthvijayan commented:
I've added one more commit.
Whenever an event is created, the `after_create :set_week` was causing one more version to be created.

![event_evil_twin](https://cloud.githubusercontent.com/assets/7537349/15788938/700e32b0-29e7-11e6-8817-cb2d11d355d0.png)


The latest commit fixes this issue.

differentreality commented:
Aren't we ignoring changes to attribute **week**?

differentreality commented on app/models/event.rb:
```diff
@@ -255,13 +255,16 @@ def generate_guid
   def set_week
-    self.week = created_at.strftime('%W')
-    save!
+    self.week = Time.zone.now.strftime('%W')
```
Why do you need this change in this PR?

nishanthvijayan replied:
The idea is to remove the save! call and set the week attribute before create

nishanthvijayan commented:
@differentreality yes we are ignoring week attribute.
I think what's happening is because of timing.Paper_trail determines object changes based on the previous version of that item.But here, when the save! inside set_week is called the create event's version has not been saved into versions table yet.So when saving the update event's version it determines that all the attributes have changes.I think this is what's happening

nishanthvijayan commented:
I've changed the fix for the  extra event version issue.
I've also implemented revert and view changes feature.When displaying changes to attributes, it'll ignore attributes where the change is from one blank type value to another Eg: if an attribute's value changes from nil to "", its ignored

![ezgif-985856866](https://cloud.githubusercontent.com/assets/7537349/15834931/692ed998-2c4b-11e6-9411-fa06280fb75e.gif)

nishanthvijayan commented:
I've changed the revert button to allow reverting single attributes of each change by means of a dropdown
![ezgif-461106150](https://cloud.githubusercontent.com/assets/7537349/15967585/eb8caddc-2f45-11e6-96da-20ba15994bc2.gif)

differentreality commented on spec/controllers/admin/versions_controller_spec.rb:
```diff
@@ -0,0 +1,89 @@
+          conference.reload
+          expect(PaperTrail::Version.last.event).to eq 'create'
+          expect(conference.program.event_types.count).to eq 2
+        end
```
I would like to know the count of event_types *prior* to the action we are testing, so that this tests makes more sense when reading it.

differentreality commented on spec/controllers/admin/versions_controller_spec.rb:
```diff
@@ -0,0 +1,89 @@
+          conference.reload
+          expect(conference.short_title).to eq 'testtitle'
+          expect(conference.description).to eq 'Some random text'
+          expect(flash[:error]).to match('Revert failed.Invalid attribute')
```
We should also check the value of title.

differentreality commented on app/controllers/admin/versions_controller.rb:
```diff
@@ -9,14 +9,35 @@ def index
+          end
+
+        else
+          flash[:error] = 'Revert failed.Invalid attribute'
```
Leave a space after the dot. There a few more places where you haven't.

differentreality commented on app/views/admin/versions/index.html.haml:
```diff
@@ -0,0 +1,48 @@
+    .col-md-9
+      %p
+        = version.whodunnit.nil? ? 'Someone (probably via the console)' : User.find(version.whodunnit).name
+
```
Name is not a mandatory field. We should show something when there is no user.name

differentreality commented on app/views/admin/versions/index.html.haml:
```diff
@@ -0,0 +1,48 @@
+          = distance_of_time_in_words(Time.now,version.created_at) + ' ago'
+
+    .col-md-3
+      .btn-group{role: "group"}
```
single quotes

differentreality commented on app/views/admin/versions/index.html.haml:
```diff
@@ -0,0 +1,48 @@
+    .col-md-3
+      .btn-group{role: "group"}
+        %a.btn.btn-success.btn-sm.show-changeset{id: "#{version.id}"} View Changes
+        %button.btn.btn-default.dropdown-toggle.btn-sm.btn-primary{"data-toggle" => "dropdown", :type => "button"}
```
single quotes, and new ruby syntax style.
You can have
`'type' => 'button' `or `type: 'button'`

differentreality commented:
Please update your 1st comment to give a short description of what you are actually implementing, what is not included, how things are supposed to work.
Eg info like:
* Only organizers can view and revert the revision history of their conferences
* Admins can view and revert the revision history for all conferences
* what can /cannot be reverted

differentreality commented on app/views/layouts/_user_menu.html.haml:
```diff
@@ -44,3 +44,7 @@
+%li
+  = link_to(admin_revision_history_path) do
+    %span.fa.fa-history
+    Revision History
```
You should only show this if a user has the appropriate access to see the page. Otherwise there will be a link which won't be accessible by the user.

differentreality commented:
The review app is erroring but locally it works (with bundle install and server restart).

differentreality commented on app/controllers/admin/versions_controller.rb:
```diff
@@ -0,0 +1,59 @@
+
+    def authorize_version
+      @version = PaperTrail::Version.find(params[:id])
+      return if current_user.has_role?(:organizer, Conference.find(@version.conference_id)) || current_user.is_admin
```
You should also change that into an ability, so that you can check if the user can do something instead of checking if the user has a role.

differentreality commented on db/migrate/20160524050538_add_conference_id_to_versions.rb:
```diff
@@ -0,0 +1,5 @@
+class AddConferenceIdToVersions < ActiveRecord::Migration
+  def change
+    add_column :versions, :conference_id, :integer
```
What about the old verion records that have nil for confernece_id? You should create a rake task to assign the conference_id for the old records too, when possible.

differentreality commented on Gemfile:
```diff
@@ -183,6 +183,10 @@ gem 'faker'
 gem 'factory_girl_rails'
 
+# for pagination
+gem 'will_paginate', '~> 3.1.0'
```
What about we just use datatable?

differentreality commented:
For adding/removing roles I would rephrase the description to sth like this:

UserX added role {organizer} to user {UserY} for conference {ConferenceX}

And the user name should also be a link.

differentreality commented on app/assets/javascripts/osem-revisionhistory.js:
```diff
@@ -0,0 +1,10 @@
+        }
+        $('#changeset-' + this.id).toggle();
+    });
+});
```
Please add an empty line at the end of the file (this is the styling we are following)

nishanthvijayan replied:
Done

differentreality commented on app/controllers/admin/versions_controller.rb:
```diff
@@ -0,0 +1,59 @@
+        flash[:notice] = 'The selected change was successfully reverted'
+
+      elsif @version.event == 'create' && %w(Conference User).include?(@version.item_type)
+        flash[:error] = 'Reverting creation of a conference is not permitted'
```
of a conference and a user?

differentreality commented on app/helpers/application_helper.rb:
```diff
@@ -276,4 +276,148 @@ def sign_in_path
+    if user_id.to_s == version.whodunnit
+      case version.event
+      when 'create' then 'registered to'
+      when 'update' then "updated #{updated_attributes(version)} of  #{name_or_username(user_id)}'s registration for"
```
let's make name_or_username(user_id) a link to that user

differentreality replied:
In this case that the registration.user is making the change we can rephrase this, as in create and destroy actions.

nishanthvijayan replied:
I didn't understand.How should it be rephrased?

differentreality replied:
Like the other 2 cases. Line 358 and 360.
Now it reads UserX updated attributes of UserX's registration.

differentreality commented on app/helpers/application_helper.rb:
```diff
@@ -276,4 +276,148 @@ def sign_in_path
+      reverse.sub(',', ' dna ').reverse
+  end
+
+  def name_or_username(user_id)
```
Does this not cover us?
https://github.com/openSUSE/osem/blob/master/app/models/user.rb#L75

nishanthvijayan replied:
It doesn't check if its blank.
[https://github.com/openSUSE/osem/blob/master/app/models/user.rb#L75](https://github.com/openSUSE/osem/blob/master/app/models/user.rb#L75) only checks if user.name is nil

differentreality replied:
Then perhaps we should fix this directly in the model and drop the helper all together.

nishanthvijayan replied:
Done

differentreality commented:
Datatable looks nice, let s put it in!

differentreality commented on lib/tasks/version.rake:
```diff
@@ -0,0 +1,23 @@
+        version.conference_id = version.item_id
+        version.save
+      elsif version.item_type == 'Event'
+        event = (version.item || version.reify || version.next.reify)
```
Even if the actual Event does not exist for some reason, the object attribute of changes has kept a record of either the conference_id (old records) or the program_id (through which we can go to conference).

That way we can ensure more (or all?) our records will get a conference_id.
So if event does not exist, search for what we want in the more hidden data ;)

nishanthvijayan replied:
The task will set the conference_id even if the event has been deleted.It fails only when if the program of that event has been deleted(which is unlikely as there is no interface to do this) as in that case there is no way to find the conference_id using the program_id

hennevogel commented on lib/tasks/version.rake:
```diff
@@ -0,0 +1,21 @@
+namespace :versions do
```
Please put this into the data namespace, not to your own

nishanthvijayan replied:
Done

differentreality commented on app/helpers/application_helper.rb:
```diff
@@ -286,4 +286,122 @@ def sign_in_path
+  end
+
+  # Recieves a PaperTrail::Version object
+  # Returns the object in its current state if its alive else the state before or after this version (after if its a create event's version)
```
Please rephrase this in a little more comprehensive way.

nishanthvijayan replied:
I've updated the comment.

differentreality commented on app/helpers/application_helper.rb:
```diff
@@ -286,4 +286,122 @@ def sign_in_path
+    elsif version.event == 'update'
+      "updated #{user.name}'s vote on"
+    else
+      "deleted #{user.name}'s vote on"
```
How do we enter this case?

differentreality commented on app/helpers/application_helper.rb:
```diff
@@ -286,4 +286,122 @@ def sign_in_path
+    if user.id.to_s == version.whodunnit && version.event == 'create'
+      'voted on'
+    elsif user.id.to_s != version.whodunnit && version.event == 'create'
+      "re-added #{user.name}'s vote on"
```
When do we enter this case?

differentreality replied:
This does not seem to appear when  reverting the action of deleting a vote. It would be nice to distinguish between a user voting and an admin deleting a vote and then reverting that deletion.
Take a look at https://osem-demo-pr-1019.herokuapp.com/admin/revision_history ID 541

nishanthvijayan replied:
Everything seems to be working correctly with Vote.
 - If User A revert the deletion of User A's vote, it'll show
   User A voted on event QWERTY
 - If User A reverts the deletion of User B's vote, it'll show:
   User A re-added User B's vote on event QWERTY

![vote](https://cloud.githubusercontent.com/assets/7537349/17066684/c9ba40d4-5065-11e6-9885-a89374d2e59a.png)

differentreality replied:
I think it is best if we use the phrasing 're-added User's X vote' in both case. Because then we can tell the difference between a user going to the proposal and voting, and a user reverting a change from the changelog.

hennevogel commented:
Crashes for me in `admin/versions/index.html.haml` with 

```
ActionView::Template::Error (undefined method `title' for #<Campaign:0x007fbdc13c6ff0>)
```

nishanthvijayan commented:
@hennevogel Is it crashing now?

differentreality commented on app/helpers/application_helper.rb:
```diff
@@ -335,4 +335,124 @@ def speaker_width(rooms)
+  end
+
+  def change_creator_link(user_id)
+    if user_id
```
It is not enough to check if a user_id was given, we need to double check the user exists before displaying a link which will otherwise fail.

differentreality commented on app/models/event.rb:
```diff
@@ -1,6 +1,6 @@
 class Event < ActiveRecord::Base
   include ActiveRecord::Transitions
-  has_paper_trail
+  has_paper_trail ignore: [:updated_at, :guid, :week], meta: { conference_id: :conference_id }
```
Should that **not** show me week in changes? (I am seeing it on updating an event)

differentreality commented:
There is a migration error when trying to create the review app, but locally it works.

differentreality commented:
It would make sense to add a revert button on the table of changes, on every line of a changed attribute, so that if you are seeing all the changes, you can just click the button (instead of going through the dropdown).

differentreality commented on lib/tasks/version.rake:
```diff
@@ -0,0 +1,21 @@
+namespace :data do
+  desc 'Sets conference_id in all pre-existing PaperTrail::Version objects'
+  task set_conference_id: :environment do
```
The name of this task is too general. This could refer to anything. 
The name of the task should tell us a bit more,eg, `set_conference_in_versions`, and it should be less generic. What if we want to set the conference_id attribute in other tables in the future... It will be confusing.

differentreality commented:
Reverting some changes is not possible (eg. user sign up). How about we don't show the revert button all together?

nishanthvijayan commented:
@differentreality So then how do we revert create/destroy of other objects? Or is that not necessary?

differentreality commented:
I mean we don't show the button only when it cannot be used!

nishanthvijayan commented:
Oh :) .I'll look into that

RichiH commented:
Instead of, or additionally to, reverting singular changes, it might also be nice to have a "make this version current" button which simply reinstates old state.

differentreality commented on spec/features/versions_spec.rb:
```diff
@@ -0,0 +1,373 @@
+
+  scenario 'display changes in user', feature: true, js: true, versioning: true do
+    sign_out
+    visit new_user_password_path
```
You have 3 different sub-scenarios to test here, you should split them up :)

differentreality commented on spec/controllers/admin/versions_controller_spec.rb:
```diff
@@ -0,0 +1,99 @@
+      end
+
+      context 'when no attribute is set' do
+        it 'reverts all changes for update events' do
```
update events is confusing, it makes me think of updating events, which is not what you are testing for here.
`update action` perhaps?

differentreality commented:
We can probably do better with our tests. I can see more than a couple of things we should rewrite. But I think it would be easier for everyone to focus on that in a different PR, after this one is merged.

differentreality commented:
I'd rather show the timestamp of the change in the beginning of the line, so that date and time don't get split up in 2 lines

differentreality commented on app/controllers/admin/versions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+            flash[:error] = 'The item is already in the state that you are trying to revert it back to'
+          else
+            @version.item[params[:attribute]] = @version.changeset[params[:attribute]][0]
+            @version.item.save!
```
What if that save does not succeed?

nishanthvijayan replied:
I've added a check to see if save fails. But adding those conditions is raising this rubocop [offence](https://travis-ci.org/openSUSE/osem/builds/145311507#L1796).

I was thinking of splitting up this controller action.One for restoring the object to the version and once action for revert a single attribute.This would make it easier to define ability and will fix the rubocop offence mentioned above

differentreality commented:
I have some older version records that for Event creation it shows a `create` record (photo 1) and an `update` record (photo 2). The update record holds the same info as the create one, same timestamps, except it adds the week. Problem is that in update record we enable the revert option, which will try to revert all attributes in nil (eg. abstract, title, event_type_id). That is of course not allowed, as we cannot have an event without a title or an abstract or an event_type_id, and the revert action throws an exception.
1) We should check if revert action is possible before showing that button. And we should authorize this a bit more in ability, so that we avoid cases like the user hitting revert and trying to make event_type_id nil.
2) You should be catching that exception to begin with and showing an error flash message in https://github.com/openSUSE/osem/pull/1019/files#diff-bc6a21dd66d0444573ff114349a614cbR27

differentreality commented on app/models/ability.rb:
```diff
@@ -166,6 +166,10 @@ def signed_in_with_organizer_role(user)
       role.resource_type == 'Conference' && (conf_ids_for_organizer.include? role.resource_id)
     end
+
+    can [:index, :revert], PaperTrail::Version do |version|
```
We should properly authorize `revert` action. Please read my latest comments on the PR about this. Then you disable the revert button based on authorization and not custom checks. Same for the revert buttons inside the table with the changes.

nishanthvijayan commented on app/views/admin/versions/_object_changes.html.haml:
```diff
@@ -9,6 +9,8 @@
           %th Previous Value
         - if version.event != 'destroy'
           %th New Value
+        - if version.event == 'update' && !version.item.nil?
```
@differentreality I want to move this check to ability.But this check should be invoked only when reverting a single attribute change.So this cannot be added to the ability with the current setup.

differentreality commented:
I think we should not be reverting the creation of an Event.
* It is not possible to perform the `destroy` action anywhere in the application, events can only be marked as `withdrawn`
* Reverting the deletion of an Event is more complicated than other models (we need to revert event.users too, which is currently not being done and things secretly break).

differentreality commented on app/assets/javascripts/osem-datatables.js:
```diff
@@ -1,6 +1,6 @@
 $(function () {
   $(document).ready(function() {
-    $('.datatable').DataTable({
+    $('.data-table').DataTable({
```
Why?

nishanthvijayan replied:
Yeah that was unnecessary.I've changed it.

nishanthvijayan replied:
Here, using 'datatable' as the classname to identify and initialize datatables is trouble in the making.
The datatable library when initializing a table as a datatable adds 'dataTable' class to the table.So if someone were to initialize a table as datatable from someplace else, like done [here](https://github.com/openSUSE/osem/blob/master/app/views/admin/volunteers/_volunteers_table.html.haml#L24) , this would cause dataTable library to raise reinitialization error alert
![reinitialize-warning](https://cloud.githubusercontent.com/assets/7537349/17050634/9d5186dc-5011-11e6-95e1-d5ffd458a8d0.png)

differentreality replied:
There is  a solution for handling that case with datatables described in the link provided by the error.

nishanthvijayan commented:
@differentreality The huge loading time is because of how we use datatable.
Describing changes requires 2-4 additional queries for each change(version).The pagination of the table is client-side and there is no server-side pagination.All versions are fetched in one go and datatable renders all of them.
So if there are 1000 versions accessible by the user, then all 1000 would be fetched and all the associated queries foir describing them will be made on loading the page instead of just fetching 10 records (& associated stuff).
Datatable has an option for [server-side processing](https://datatables.net/release-datatables/examples/data_sources/server_side.html)

> With server-side processing enabled, all paging, searching, ordering actions that DataTables performs are handed off to the server...As such, each draw of the table will result in a new Ajax request being made to get the required data.

But in the exmaple provided in their docs, the data to displayed is simple, but in our case its complex html (buttons and such).Also another problem with this approach is that we'll have to implement pagination (and sorting and searching if needed) in the server-side.I don't think its worth the effort.My suggestion is that we go back to using will-paginate and drop Datatables.

differentreality commented:
@nishanthvijayan we should investigate on how we can keep sorting and search (which should be applied on the whole set of objects, not just the ones visible on the page). But I would leave this for a new PR.

nishanthvijayan commented:
Sorting already works for the whole set of version objects.

differentreality commented:
@nishanthvijayan this has been a lot and very good work :) 
You can go ahead and squash commits and make sure you rebase. I have added a comment about some phrasing, please also fix this.

I will make a final review during the next few days, and hopefully we can get this in.

differentreality commented:
@nishanthvijayan please rebase again to fix travis failure

nishanthvijayan commented:
@differentreality In the last commit, I've reintroduced datatable and have attempted to implement serverside processing similar to #1117

nishanthvijayan commented:
@differentreality I've removed the last commit and all the will_paginate stuff.It uses datatables now.

differentreality commented:
@nishanthvijayan please rebase all your PRs so that we can start merging.

nishanthvijayan commented:
@differentreality Done

differentreality commented:
:+1:
